README.md
[![Build Status](https://travis-ci.org/sertansenturk/alignednotemodel.svg?branch=master)](https://travis-ci.org/sertansenturk/alignednotemodel) [![Code Climate](https://codeclimate.com/github/sertansenturk/alignednotemodel/badges/gpa.svg)](https://codeclimate.com/github/sertansenturk/alignednotemodel)

# alignednotemodels

Python tools to compute note models from note-level audio-score alignment.

Currently the algorithm computes the stable pitch and a pitchdistribution of each aligned note.

Usage
=======

```python
from alignednotemodel.alignednotemodel import AlignedNoteModel
alignedNoteModel = AlignedNoteModel(kernel_width=7.5, step_size=7.5, pitch_threshold=50)

note_models, pitch_distribution, new_tonic = alignedNoteModel.get_models(pitch, aligned_notes,
    tonic_symbol)
```

Instantiation parameters are:
```python
# kernel_width    : The width of the Gaussian kernel used to compute the pitch distribution 
#                   (default: 7.5 cent ~ 1/3 Hc)
# step_size       : The step size between each bin of the pitch distribution (default: 7.5 cent 
#                   ~ 1/3 Hc)
# pitch_threshold : Max cent difference for two pitch calues to be considered close. Used in
#                   stable pitch computation (default: 50 cent, a quarter tone)
```

The inputs for the get_models method are:
```python
# pitch 		  :	an n-by-2 matrix, where the values in the first column are 
#					the timestamps and the values in the second column are frequency 
#					values
# aligned_notes	  :	the list of aligned notes. This is read from the alignedNotes.json 
#					output from the fragmentLinker (https://github.com/sertansenturk/fragmentLinker) 
#                   repository 
# tonic_symbol	  : The tonic symbol in the symbTr format (e.g. B4b1)
```

The outputs are:
```python
# note_models        : The model for each note symbol
# pitch_distribution : The pitch distribution computed from the pitch input
# new_tonic		     : The updated tonic according to the note model of the tonic symbol
```

Please refer to [note_model.ipynb](https://github.com/sertansenturk/alignednotemodel/blob/master/note_model.ipynb) for an interactive demo.

Installation
============

If you want to install the repository, it is recommended to install the package and dependencies into a virtualenv. In the terminal, do the following:

    virtualenv env
    source env/bin/activate
    python setup.py install

If you want to be able to edit files and have the changes be reflected, then
install the repository like this instead

    pip install -e .

The algorithm uses several modules in Essentia. Follow the [instructions](essentia.upf.edu/documentation/installing.html) to install the library.

Now you can install the rest of the dependencies:

    pip install -r requirements

Authors
-------
Sertan Şentürk
contact@sertansenturk.com

Reference
-------
Thesis


Changelog.md
#### alignednotemodel v1.1.1
 - Updated PitchDistribution requirements (morty)

#### alignednotemodel v1.1
 - Refactorings to improve code quality, readability and functionality

#### alignednotemodel v1.0
 - First public release


alignednotemodel/AlignedNoteModel.py
from modetonicestimation.PitchDistribution import PitchDistribution
from modetonicestimation.Converter import Converter
import numpy as np
import matplotlib.pyplot as plt
from copy import deepcopy
import os
import json


class AlignedNoteModel(object):
    def __init__(self, kernel_width=7.5, step_size=7.5, pitch_threshold=50):
        self.kernel_width = kernel_width
        self.step_size = step_size
        self.pitch_threshold = pitch_threshold  # max threshold for two
        # pitches to be considered close. Used in stable pitch computation

    def get_models(self, pitch, alignednotes, tonic_symbol):
        note_file = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                 'data', 'note_dict.json')
        note_dict = json.load(open(note_file, 'r'))

        pitch = np.array(pitch)
        alignednotes_ext = deepcopy(alignednotes)

        note_names = set(an['Symbol'] for an in alignednotes_ext)
        note_models = dict((
            nn, {'notes': [], 'distribution': [], 'stable_pitch': [],
                 'performed_interval': [],
                 'theoretical_interval': {
                     'Value': (note_dict[nn]['Value'] -
                               note_dict[tonic_symbol]['Value']),
                     'Unit': 'cent'},
                 'theoretical_pitch': []}) for nn in note_names)

        # compute note trajectories and add to each model
        for an in alignednotes_ext:
            if not an['Interval'][0] == an['Interval'][1]:  # not aligned
                trajectory = np.vstack(
                    p for p in pitch
                    if an['Interval'][0] <= p[0] <= an['Interval'][1])
                notetemp = dict(an)
                notetemp['PitchTrajectory'] = trajectory

                note_models[an['Symbol']]['notes'].append(notetemp)

        # remove models without any aligned note
        for key in note_models.keys():
            if not note_models[key]['notes']:
                note_models.pop(key, None)

        # update the tonic frequency temporarily
        # NOTE: extremely unlikely but this value might shift to the next bin
        # in the note model computation. Hence we don't assign it to the
        # final tonic value.
        tonic_trajectories = [nn['PitchTrajectory'][:, 1]
                              for nn in note_models[tonic_symbol]['notes']]
        temp_tonic_freq = self._get_stablepitch_distribution(
            tonic_trajectories,
            note_models[tonic_symbol]['theoretical_interval']['Value'])[0]

        # compute the histogram for each model
        for nm in note_models.values():
            peak_freq, distribution = self._get_stablepitch_distribution(
                [nn['PitchTrajectory'][:, 1] for nn in nm['notes']],
                nm['theoretical_interval']['Value'],
                ref_freq=temp_tonic_freq)

            nm['stable_pitch'] = {'Value': peak_freq, 'Unit': 'Hz'}
            nm['distribution'] = distribution

        # update the new tonic frequency
        newtonic = {'alignment': {
            'Value': note_models[tonic_symbol]['stable_pitch']['Value'],
            'Unit': 'Hz', 'Symbol': tonic_symbol, 'Method': 'alignedNoteModel',
            'OctaveWrapped': False, 'Citation': 'SenturkPhDThesis',
            'Procedure': 'Tonic identified from the tonic note model obtained '
                         'from audio-score alignment'}}

        # get the distances wrt tonic
        for nm in note_models.values():
            interval = Converter.hz_to_cent(nm['stable_pitch']['Value'],
                                            newtonic['alignment']['Value'])
            nm['performed_interval'] = {'Value': interval, 'Unit': 'cent'}

            theo_pitch = Converter.cent_to_hz(
                nm['theoretical_interval']['Value'],
                newtonic['alignment']['Value'])
            nm['theoretical_pitch'] = {'Value': theo_pitch, 'Unit': 'Hz'}

        # compute the complete histogram without normalization
        recording_distribution = PitchDistribution.from_hz_pitch(
            pitch, ref_freq=temp_tonic_freq, smooth_factor=self.kernel_width,
            step_size=self.step_size, norm_type=None)
        recording_distribution.cent_to_hz()

        # normalize all the distributions
        recording_distribution, note_models = self._normalize_distributions(
            recording_distribution, note_models)

        return note_models, recording_distribution, newtonic

    def _get_stablepitch_distribution(self, note_trajectories,
                                      theoretical_interval, ref_freq=None):
        temp_pitch_vals = np.hstack(nn for nn in note_trajectories)

        # useful to keep the bins coinciding with a desired value,
        # e.g. tonic frequency
        if ref_freq is None:
            ref_freq = self._get_median_pitch(temp_pitch_vals)

        distribution = PitchDistribution.from_hz_pitch(
            temp_pitch_vals, ref_freq=ref_freq,
            smooth_factor=self.kernel_width, step_size=self.step_size,
            norm_type=None)

        # get the stable pitch as the highest peaks among the peaks close to
        # the theoretical pitch TODO
        peaks = distribution.detect_peaks()
        peak_bins = distribution.bins[peaks[0]]
        peak_vals = distribution.vals[peaks[0]]

        try:
            cand_bool = (abs(peak_bins - theoretical_interval) <
                         self.pitch_threshold)
            stable_pitch_cand = peak_bins[cand_bool]
            cand_occr = peak_vals[cand_bool]

            peak_cent = stable_pitch_cand[np.argmax(cand_occr)]

            # convert to hz scale
            peak_freq = Converter.cent_to_hz(peak_cent, ref_freq)
        except ValueError:  # no stable pitch in the vicinity, probably a
            # misalignment
            peak_freq = None

        # convert to hz scale
        distribution.cent_to_hz()

        return peak_freq, distribution

    @staticmethod
    def _get_median_pitch(pitch):
        if pitch.ndim > 1:
            pitch = pitch[:, 1]

        # filter the nan, inf and inaudible
        pitch = pitch[~np.isnan(pitch)]
        pitch = pitch[~np.isinf(pitch)]
        pitch = pitch[pitch >= 20.0]

        return np.median(pitch)

    @staticmethod
    def _normalize_distributions(recording_distribution, note_models):
        dist_norm = deepcopy(recording_distribution)

        # sum normalization on the audio recording
        dist_norm.normalize(norm_type='sum')

        # compute the normalization factor
        ini_max_val = max(recording_distribution.vals)
        fin_max_val = max(dist_norm.vals)
        norm_factor = fin_max_val / ini_max_val

        for nm in note_models.values():
            nm['distribution'].vals *= norm_factor

        return dist_norm, note_models

    @staticmethod
    def to_json(note_models, json_path=None):
        # conversions for json serialization
        note_models_ser = deepcopy(note_models)
        for nm in note_models_ser.values():
            # convert the distribution object to dict
            nm['distribution'] = nm['distribution'].to_dict()

            # convert the pitchtrajectories to lists from numpy arrays
            for nn in nm['notes']:
                nn['PitchTrajectory'] = nn['PitchTrajectory'].tolist()

        if json_path is None:
            return json.dumps(note_models_ser, indent=4)
        else:
            json.dump(note_models_ser, open(json_path, 'w'), indent=4)

    @staticmethod
    def plot(note_models, pitch_distribution, alignednotes, pitch):
        pitch = np.array(pitch)

        fig, (ax1, ax2) = plt.subplots(1, 2, sharey=True)
        ax1.plot(pitch[:, 0], pitch[:, 1], 'g', label='Pitch', alpha=0.7)
        fig.subplots_adjust(wspace=0)
        ax1.set_xlabel('Time (sec)')
        ax1.yaxis.grid(True)

        for note in alignednotes:
            ax1.plot(
                note['Interval'], [note['PerformedPitch']['Value'],
                                   note['PerformedPitch']['Value']],
                'r', alpha=0.4, linewidth=4)

        ax2.plot(pitch_distribution.vals, pitch_distribution.bins, '-.',
                 color='#000000', alpha=0.9)
        for key in note_models.keys():
            ax2.plot(note_models[key]['distribution'].vals,
                     note_models[key]['distribution'].bins, label=key)

        ax2.set_yticks(
            [nm['stable_pitch']['Value'] for nm in note_models.values()])
        ax2.set_yticklabels(
            [key + ', ' + "%.1f" % val['stable_pitch']['Value'] +
             ' Hz, ' + "%.1f" % val['performed_interval']['Value'] + ' cents'
             for key, val in note_models.iteritems()])
        ax2.axis('off')
        ax2.set_ylim([np.min(pitch_distribution.bins),
                      np.max(pitch_distribution.bins)])
        ax2.yaxis.grid(True)

        ax2.set_xticklabels([])
        ax2.spines['top'].set_visible(False)
        ax2.spines['right'].set_visible(False)
        ax2.spines['bottom'].set_visible(False)
        ax2.spines['left'].set_visible(False)

        return fig, (ax1, ax2)


alignednotemodel/PitchDistribution.py
# -*- coding: utf-8 -*-
import essentia
import essentia.standard as std
import numpy as np
import json
import os
from scipy.integrate import simps
from scipy.stats import norm
import matplotlib.pyplot as plt

def load(fname):
	"""-------------------------------------------------------------------------
	Loads a PitchDistribution object from JSON file.
	----------------------------------------------------------------------------
	fname    : The filename of the JSON file
	-------------------------------------------------------------------------"""
	with open(fname) as f:
		dist = json.load(f)

	return PitchDistribution(np.array(dist[0]['bins']), np.array(dist[0]['vals']),
		                     kernel_width=dist[0]['kernel_width'], ref_freq=dist[0]['ref_freq'])

class PitchDistribution:
	
	def __init__(self, pd_bins, pd_vals, kernel_width=7.5, ref_freq=440):
		"""------------------------------------------------------------------------
		The main data structure that wraps all the relevant information about a 
		pitch distribution.
		---------------------------------------------------------------------------
		pd_bins      : Bins of the pitch distribution. It is a 1-D list of equally
		               spaced monotonically increasing frequency values.
		pd_vals      : Values of the pitch distribution
		kernel_width : The std. deviation of the Gaussian kernel. See generate_pd()
		               of ModeFunctions for more detail.
		ref_freq     : Reference frequency that is used while generating the 
		               distribution. If the tonic of a recording is annotated,
		               this is variable that stores it.
		-------------------------------------------------------------------------"""
		self.bins = pd_bins
		self.vals = pd_vals
		self.kernel_width = kernel_width
		self.ref_freq = ref_freq

		### Due to the floating point issues in Python, the step_size might not be
		### exactly equal to (for example) 7.5, but 7.4999... In such cases the 
		### bin generation of pitch distributions include 1200 cents too and chaos
		### reigns. We fix it here.
		temp_ss = self.bins[1] - self.bins[0]
		self.step_size = (temp_ss if temp_ss == (round(temp_ss * 10) / 10) 
			else round(temp_ss * 10) / 10)

	def save(self, fname, save_dir='./'):
		"""-------------------------------------------------------------------------
		Saves the PitchDistribution object to a JSON file.
		----------------------------------------------------------------------------
		fname    : The name of the JSON file to be created.
		save_dir : Pathway of where the JSON would saved
		-------------------------------------------------------------------------"""
		dist_json = [{'bins':self.bins.tolist(), 'vals':self.vals.tolist(),
		              'kernel_width':self.kernel_width, 'ref_freq':self.ref_freq, 
		              'step_size':self.step_size}]
		with open(os.path.join(save_dir, fname), 'w') as f:
			json.dump(dist_json, f, indent=2)
			f.close()

	def is_pcd(self):
		"""-------------------------------------------------------------------------
		The boolean flag of whethwe the instance is PCD or not.
		-------------------------------------------------------------------------"""
		return (max(self.bins) == (1200 - self.step_size) and min(self.bins) == 0)

	def detect_peaks(self):
		"""-------------------------------------------------------------------------
		Finds the peak indices of the distribution. These are treated as tonic
		candidates in higher order functions.
		-------------------------------------------------------------------------"""
		# Peak detection is handled by Essentia
		detector = std.PeakDetection()
		peak_bins, peak_vals = detector(essentia.array(self.vals))

		# Essentia normalizes the positions to 1, they are converted here
		# to actual index values to be used in bins.
		peak_idxs = [round(bn * (len(self.bins) - 1)) for bn in peak_bins]
		if(peak_idxs[0] == 0):
			peak_idxs = np.delete(peak_idxs, [len(peak_idxs) - 1])
			peak_vals = np.delete(peak_vals, [len(peak_vals) - 1])
		return peak_idxs, peak_vals

	def shift(self, shift_idx):
		"""-------------------------------------------------------------------------
		Shifts the distribution by the given number of samples
		----------------------------------------------------------------------------
		shift_idx : The number of samples that the distribution is tı be shifted
		-------------------------------------------------------------------------"""
		# If the shift index is non-zero, do the shifting procedure
		if(shift_idx):
			
			# If distribution is a PCD, we do a circular shift
			if self.is_pcd():
				shifted_vals = np.concatenate((self.vals[shift_idx:], self.vals[:shift_idx]))
			
			# If distribution is a PD, it just shifts the values. In this case,
			# pd_zero_pad() of ModeFunctions is always applied beforehand to make
			# sure that no non-zero values are dropped.
			else:
				
				# Shift towards left
				if(shift_idx > 0):
					shifted_vals = np.concatenate((self.vals[shift_idx:], np.zeros(shift_idx)))
				
				# Shift towards right
				else: 
					shifted_vals = np.concatenate((np.zeros(abs(shift_idx)), self.vals[:shift_idx]))

			return PitchDistribution(self.bins, shifted_vals, kernel_width=self.kernel_width,
				ref_freq=self.ref_freq)
		
		# If a zero sample shift is requested, a copy of the original distribution
		# is returned
		else:
			return PitchDistribution(self.bins, self.vals, kernel_width=self.kernel_width,
				ref_freq=self.ref_freq)

	def plot(self):
		plt.plot(self.bins, self.vals)

def generate_pd(cent_track, ref_freq=440, kernel_width=7.5, step_size=7.5):
	"""-------------------------------------------------------------------------
	Given the pitch track in the unit of cents, generates the Pitch Distribution
	of it. the pitch track from a text file. 0th column is the time-stamps and
	1st column is the corresponding frequency values.
	----------------------------------------------------------------------------
	cent_track:     1-D array of frequency values in cents.
	ref_freq:       Reference frequency used while converting Hz values to cents.
	                This number isn't used in the computations, but is to be
	                recorded in the PitchDistribution object.
	kernel_width:  The standard deviation of the gaussian kernel, used in Kernel
	                Density Estimation. If 0, a histogram is given
	step_size:        The step size of the Pitch Distribution bins.
	-------------------------------------------------------------------------"""

	### Some extra interval is added to the beginning and end since the
	### superposed Gaussian for kernel_width would introduce some tails in the
	### ends. These vanish after 3 sigmas(=kernel_width).

	### The limits are also quantized to be a multiple of chosen step-size
	### kernel_width = standard deviation fo the gaussian kernel

	### TODO: filter out the NaN, -infinity and +infinity from the pitch track

	# Finds the endpoints of the histogram edges. Histogram bins will be
	# generated as the midpoints of these edges. 
	min_edge = min(cent_track) - (step_size / 2.0) - kernel_width*3
	max_edge = max(cent_track) + (step_size / 2.0) + kernel_width*3
	pd_edges = np.concatenate([np.arange(-step_size/2.0, min_edge, -step_size)[::-1],
	                           np.arange(step_size/2.0, max_edge, step_size)])

	# An exceptional case is when min_bin and max_bin are both positive or negative
	# In this case, remove unnecessary bins

	if min_edge > 0:
		pd_edges = [e for e in pd_edges if e >= min_edge]

	if max_edge < 0:
		pd_edges = [e for e in pd_edges if e <= max_edge]
	
	# Generates the histogram and bins (i.e. the midpoints of edges)
	pd_vals, pd_edges = np.histogram(cent_track, bins=pd_edges, density=True)
	pd_bins = np.convolve(pd_edges, [0.5,0.5])[1:-1]

	if kernel_width > 0: # kernel density estimation (approximated)
		# smooth the histogram
		normal_dist = norm(loc = 0, scale = kernel_width)
		xn = np.concatenate([np.arange(0, - 5 * kernel_width, -step_size)[::-1], 
		    np.arange(step_size, 5 * kernel_width, step_size)])
		sampled_norm = normal_dist.pdf(xn)

		extra_num_bins = len(sampled_norm)/2 # convolution generates tails
		pd_vals = np.convolve(pd_vals, sampled_norm)[extra_num_bins:-extra_num_bins]

		# normalize the area under the curve
		area = simps(pd_vals, dx=step_size)
		pd_vals = pd_vals/area

	# Sanity check. If the histogram bins and vals lengths are different, we
	# are in trouble. This is an important assumption of higher level functions.
	if(len(pd_bins) != len(pd_vals)):
		raise ValueError('Lengths of bins and Vals are different')

	# Initializes the PitchDistribution object and returns it.
	return PitchDistribution(pd_bins, pd_vals, kernel_width=kernel_width, ref_freq=ref_freq)

def generate_pcd(pd):
	"""-------------------------------------------------------------------------
	Given the pitch distribution of a recording, generates its pitch class
	distribution, by octave wrapping.
	----------------------------------------------------------------------------
	pd: PitchDistribution object. Its attributes include everything we need
	-------------------------------------------------------------------------"""

	# Initializations
	pcd_bins = np.arange(0, 1200, pd.step_size)
	pcd_vals = np.zeros(len(pcd_bins))

	# Octave wrapping
	for k in range(len(pd.bins)):
		idx = int((pd.bins[k] % 1200) / pd.step_size)
		idx = idx if idx != 160 else 0
		pcd_vals[idx] += pd.vals[k]

	# Initializes the PitchDistribution object and returns it.
	return (pcd_bins, pcd_val)


def hz_to_cent(hz_track, ref_freq):
	"""-------------------------------------------------------------------------
	Converts an array of Hertz values into cents.
	----------------------------------------------------------------------------
	hz_track : The 1-D array of Hertz values
	ref_freq	: Reference frequency for cent conversion
	-------------------------------------------------------------------------"""
	if type(hz_track)==list:
		hz_track = np.array(hz_track)
	else:
		hz_track = np.array([hz_track])

	# The 0 Hz values are removed, not only because they are meaningless,
	# but also logarithm of 0 is problematic.
	return np.log2(hz_track[hz_track>0] / ref_freq) * 1200.0


def cent_to_hz(cent_track, ref_freq):
	"""-------------------------------------------------------------------------
	Converts an array of cent values into Hertz.
	----------------------------------------------------------------------------
	cent_track  : The 1-D array of cent values
	ref_freq	: Reference frequency for cent conversion
	-------------------------------------------------------------------------"""
	cent_track = np.array(cent_track)

	return 2 ** (cent_track / 1200.0) * ref_freq


alignednotemodel/alignednotemodel.py
from morty.pitchdistribution import PitchDistribution
from morty.converter import Converter
import numpy as np
import matplotlib.pyplot as plt
from copy import deepcopy
import os
import json
import logging
logging.basicConfig(level=logging.INFO)


class AlignedNoteModel(object):
    def __init__(self, kernel_width=7.5, step_size=7.5, pitch_threshold=50):
        self.kernel_width = kernel_width
        self.step_size = step_size
        self.pitch_threshold = pitch_threshold  # max threshold for two
        # pitches to be considered close. Used in stable pitch computation

    def get_models(self, pitch, alignednotes, tonic_symbol):
        note_file = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                                 'data', 'note_dict.json')
        note_dict = json.load(open(note_file, 'r'))

        pitch = np.array(pitch)
        alignednotes_ext = deepcopy(alignednotes)

        note_names = set(an['Symbol'] for an in alignednotes_ext)
        note_models = {}
        for nn in note_names:
            try:
                note_models[nn] = {
                    'notes': [], 'distribution': [], 'stable_pitch': [],
                    'performed_interval': [], 'theoretical_interval': {
                        'Value': (note_dict[nn]['Value'] -
                                  note_dict[tonic_symbol]['Value']),
                        'Unit': 'cent'}, 'theoretical_pitch': []}
            except KeyError:
                logging.warning(
                    u"The note {0:s} is not in the note_dict.".format(nn))

        # compute note trajectories and add to each model
        self._distribute_pitch_trajectories(alignednotes_ext, note_models,
                                            pitch)

        # remove models without any aligned note
        self._remove_unaligned_notes(note_models)

        # update the tonic frequency temporarily
        # NOTE: extremely unlikely but this value might shift to the next bin
        # in the note model computation. Hence we don't assign it to the
        # final tonic value.
        tonic_trajectories = [nn['PitchTrajectory'][:, 1]
                              for nn in note_models[tonic_symbol]['notes']]
        temp_tonic_freq = self._get_stablepitch_distribution(
            tonic_trajectories,
            note_models[tonic_symbol]['theoretical_interval']['Value'])[0]

        # compute the histogram for each model
        self._get_note_histogram(note_models, temp_tonic_freq)

        # update the new tonic frequency
        newtonic = {'alignment': {
            'Value': note_models[tonic_symbol]['stable_pitch']['Value'],
            'Unit': 'Hz', 'Symbol': tonic_symbol, 'Method': 'alignedNoteModel',
            'OctaveWrapped': False, 'Citation': 'SenturkPhDThesis',
            'Procedure': 'Tonic identified from the tonic note model obtained '
                         'from audio-score alignment'}}

        # get the distances wrt tonic
        self._get_tunings(newtonic, note_models)

        # compute the complete histogram without normalization
        recording_distribution = PitchDistribution.from_hz_pitch(
            pitch, ref_freq=temp_tonic_freq, kernel_width=self.kernel_width,
            step_size=self.step_size, norm_type=None)
        recording_distribution.cent_to_hz()

        # normalize all the distributions
        recording_distribution, note_models = self._normalize_distributions(
            recording_distribution, note_models)

        return note_models, recording_distribution, newtonic

    def _get_note_histogram(self, note_models, temp_tonic_freq):
        for nm in note_models.values():
            peak_freq, distribution = self._get_stablepitch_distribution(
                [nn['PitchTrajectory'][:, 1] for nn in nm['notes']],
                nm['theoretical_interval']['Value'],
                ref_freq=temp_tonic_freq)

            nm['stable_pitch'] = {'Value': peak_freq, 'Unit': 'Hz'}
            nm['distribution'] = distribution

    @staticmethod
    def _get_tunings(newtonic, note_models):
        for nm in note_models.values():
            interval = Converter.hz_to_cent(nm['stable_pitch']['Value'],
                                            newtonic['alignment']['Value'])
            nm['performed_interval'] = {'Value': interval, 'Unit': 'cent'}

            theo_pitch = Converter.cent_to_hz(
                nm['theoretical_interval']['Value'],
                newtonic['alignment']['Value'])
            nm['theoretical_pitch'] = {'Value': theo_pitch, 'Unit': 'Hz'}

    @staticmethod
    def _remove_unaligned_notes(note_models):
        for key in note_models.keys():
            if not note_models[key]['notes']:
                note_models.pop(key, None)

    @staticmethod
    def _distribute_pitch_trajectories(alignednotes_ext, note_models, pitch):
        for an in alignednotes_ext:
            if not an['Interval'][0] == an['Interval'][1]:  # not aligned
                trajectory = np.vstack(
                    p for p in pitch
                    if an['Interval'][0] <= p[0] <= an['Interval'][1])
                notetemp = dict(an)
                notetemp['PitchTrajectory'] = trajectory

                try:
                    note_models[an['Symbol']]['notes'].append(notetemp)
                except KeyError:
                    logging.info(u"The note {0:s} is not in the note_dict."
                                 u"".format(an['Symbol']))

    def _get_stablepitch_distribution(self, note_trajectories,
                                      theoretical_interval, ref_freq=None):
        temp_pitch_vals = np.hstack(nn for nn in note_trajectories)

        # useful to keep the bins coinciding with a desired value,
        # e.g. tonic frequency
        if ref_freq is None:
            ref_freq = self._get_median_pitch(temp_pitch_vals)

        distribution = PitchDistribution.from_hz_pitch(
            temp_pitch_vals, ref_freq=ref_freq,
            kernel_width=self.kernel_width, step_size=self.step_size,
            norm_type=None)

        # get the stable pitch as the highest peaks among the peaks close to
        # the theoretical pitch TODO
        peaks = distribution.detect_peaks()
        peak_bins = distribution.bins[peaks[0]]
        peak_vals = distribution.vals[peaks[0]]

        try:
            cand_bool = (abs(peak_bins - theoretical_interval) <
                         self.pitch_threshold)
            stable_pitch_cand = peak_bins[cand_bool]
            cand_occr = peak_vals[cand_bool]

            peak_cent = stable_pitch_cand[np.argmax(cand_occr)]

            # convert to hz scale
            peak_freq = Converter.cent_to_hz(peak_cent, ref_freq)
        except ValueError:  # no stable pitch in the vicinity, probably a
            # misalignment
            peak_freq = None

        # convert to hz scale
        distribution.cent_to_hz()

        return peak_freq, distribution

    @staticmethod
    def _get_median_pitch(pitch):
        if pitch.ndim > 1:
            pitch = pitch[:, 1]

        # filter the nan, inf and inaudible
        pitch = pitch[~np.isnan(pitch)]
        pitch = pitch[~np.isinf(pitch)]
        pitch = pitch[pitch >= 20.0]

        return np.median(pitch)

    @staticmethod
    def _normalize_distributions(recording_distribution, note_models):
        dist_norm = deepcopy(recording_distribution)

        # sum normalization on the audio recording
        dist_norm.normalize(norm_type='sum')

        # compute the normalization factor
        ini_max_val = max(recording_distribution.vals)
        fin_max_val = max(dist_norm.vals)
        norm_factor = fin_max_val / ini_max_val

        for nm in note_models.values():
            nm['distribution'].vals *= norm_factor

        return dist_norm, note_models

    @staticmethod
    def to_json(note_models, json_path=None):
        # conversions for json serialization
        note_models_ser = deepcopy(note_models)
        for nm in note_models_ser.values():
            # convert the distribution object to dict
            nm['distribution'] = nm['distribution'].to_dict()

            # convert the pitchtrajectories to lists from numpy arrays
            for nn in nm['notes']:
                nn['PitchTrajectory'] = nn['PitchTrajectory'].tolist()

        if json_path is None:
            return json.dumps(note_models_ser, indent=4)
        else:
            json.dump(note_models_ser, open(json_path, 'w'), indent=4)

    @staticmethod
    def plot(note_models, pitch_distribution, alignednotes, pitch):
        pitch = np.array(pitch)

        fig, (ax1, ax2) = plt.subplots(1, 2, sharey=True)
        ax1.plot(pitch[:, 0], pitch[:, 1], 'g', label='Pitch', alpha=0.7)
        fig.subplots_adjust(wspace=0)
        ax1.set_xlabel('Time (sec)')
        ax1.yaxis.grid(True)

        for note in alignednotes:
            ax1.plot(
                note['Interval'], [note['PerformedPitch']['Value'],
                                   note['PerformedPitch']['Value']],
                'r', alpha=0.4, linewidth=4)

        ax2.plot(pitch_distribution.vals, pitch_distribution.bins, '-.',
                 color='#000000', alpha=0.9)
        for key in note_models.keys():
            ax2.plot(note_models[key]['distribution'].vals,
                     note_models[key]['distribution'].bins, label=key)

        ax2.set_yticks(
            [nm['stable_pitch']['Value'] for nm in note_models.values()])
        ax2.set_yticklabels(
            [key + ', ' + "%.1f" % val['stable_pitch']['Value'] +
             ' Hz, ' + "%.1f" % val['performed_interval']['Value'] + ' cents'
             for key, val in note_models.iteritems()])
        ax2.axis('off')
        ax2.set_ylim([np.min(pitch_distribution.bins),
                      np.max(pitch_distribution.bins)])
        ax2.yaxis.grid(True)

        ax2.set_xticklabels([])
        ax2.spines['top'].set_visible(False)
        ax2.spines['right'].set_visible(False)
        ax2.spines['bottom'].set_visible(False)
        ax2.spines['left'].set_visible(False)

        return fig, (ax1, ax2)
